tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/UJsonTraceReader.scala

```diff
@@ -0,0 +1,95 @@
+package at.forsyte.apalache.tla.tracee
+
+import at.forsyte.apalache.infra.passes.options.SourceOption
+import at.forsyte.apalache.infra.passes.options.SourceOption.FileSource
+import at.forsyte.apalache.io.json.impl.{UJsonRep, UJsonScalaFactory}
+import at.forsyte.apalache.infra.passes.options.SourceOption._
+import at.forsyte.apalache.io.json.JsonDeserializationError
+import at.forsyte.apalache.io.lir.TypeTagReader
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.oper.{TlaBoolOper, TlaOper}
+import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx, TlaOperDecl}
+import at.forsyte.apalache.tla.typecomp.json.UJsonToTlaViaBuilder
+
+import scala.annotation.unused
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Reads an ITF-format or a TlaJSON-format trace into a TLAIR trace, with an intermediate UJson representation.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UJsonTraceReader(sourceStoreOpt: Option[SourceStore], tagReader: TypeTagReader) extends TraceReader[UJsonRep] {
+  private val builder = new UJsonToTlaViaBuilder(sourceStoreOpt)(tagReader)
+  override def read(source: SourceOption): UJsonRep = {
+    val readable: ujson.Readable = source match {
+      case FileSource(f, Format.Json) => f
+      case _                          => throw new IllegalArgumentException("File provided with --trace is not a JSON.")
```

Comment:
This should also support `StringSource` I think, no?

Also, imo, probably not a great idea to mention the CLI flag in this error. If the flag name changes this will be out of date, and users could load this through a config or RPC call, or this class could be used for other purposes, no through your new CLI endpoint. So probably better to have something like:

```scala
      case _                          => throw new IllegalArgumentException("Tried to load trace from a non-JSON source.")
```

If you need give more specific cli input based input, you can catch this exception (or, better, maybe this should be `Try`  ;P) and extend the error.

Reply:
Why should this support reading from files but not reading from strings if it is going to suppoort `SourceOption` at all? If you change the method, so it only takes `ujson` that seems consistent, but if you are accepting `SourceOption` here, I see no reason it shouldn't also support `StringSource`. Adding support should also be a one line addition iiuc.